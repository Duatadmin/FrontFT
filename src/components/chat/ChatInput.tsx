import React, { useState, useRef, useEffect, KeyboardEvent, memo, useCallback } from 'react';
import { AnimatePresence, motion } from "motion/react";

import { DashboardButton } from '../chat/DashboardButton'; // Adjusted path
import VoiceWidget from '../VoiceWidget'; // Import original VoiceWidget
import VoiceTicker, { ISepiaVoiceRecorder } from './VoiceTicker'; // Import original VoiceTicker
// import { VoiceModeToggle } from '../chat/VoiceModeToggle';   // Adjusted path
// import { WalkieTalkieButton } from '../chat/WalkieTalkieButton';// Adjusted path
import { LottieSendButton } from './LottieSendButton';

import { useVoicePlayback } from '../../hooks/useVoicePlayback';

interface ChatInputProps {
  onSendMessage: (message: string) => void;
  isLoading: boolean;
  onTtsPlaybackStart?: () => void;
  onTtsPlaybackEnd?: () => void;
}

const ChatInput: React.FC<ChatInputProps> = memo(({ 
  onSendMessage, 
  isLoading,
}) => {
  const [voiceWidgetStatus, setVoiceWidgetStatus] = useState<string>('idle'); // To track VoiceWidget state
    const [isDemoActive, /* setIsDemoActive */] = useState(false); // For testing the ticker animation
  const voiceTickerRecorderRef = useRef<ISepiaVoiceRecorder>({ onResamplerData: undefined });
  const [isSending, setIsSending] = useState(false); // Local state to prevent race conditions
  // console.log('[ChatInput] Initializing: isSending = false, isLoading =', isLoading); // Initial state log
  const isSendingRef = useRef(false); // Ref for immediate synchronous check
  const [inputValue, setInputValue] = useState('');
  const [hasInteracted, setHasInteracted] = useState(false);
  const textareaRef = useRef<HTMLTextAreaElement>(null);

  // ---------------- Animated Placeholder ----------------
  const placeholders = [
    "Let's start a new training session!",
    "I am borred with my current training plan. Lets create a new one!",
    "Tell me a joke!",
    "Ask anything...",
  ];
  const [currentPlaceholder, setCurrentPlaceholder] = useState(0);
  useEffect(() => {
    const id = setInterval(() => {
      setCurrentPlaceholder((prev) => (prev + 1) % placeholders.length);
    }, 3000);
    return () => clearInterval(id);
  }, []);
  // ------------------------------------------------------
  const chatInputId = 'chat-input'; 


  // const chatInputId = 'chat-input'; // This was the duplicate, original is near line 25

  useVoicePlayback(); // voiceEnabled and toggleVoice are not used after removing VoiceModeToggle
  
  useEffect(() => {
    const initVoice = async () => {
      try {
      } catch (error) {
        console.error('Failed to initialize voice module in chat input:', error);
      }
    };
    
    initVoice();
  }, []);
  
  useEffect(() => {
    const finalTranscriptEl = document.getElementById(`${chatInputId}-final`);
    const interimTranscriptEl = document.getElementById(`${chatInputId}-interim`);

    if (!finalTranscriptEl || !interimTranscriptEl) {
      const transcriptContainer = document.createElement('div');
      transcriptContainer.style.display = 'none';
      transcriptContainer.id = `${chatInputId}-transcript-container`;
      
      const interimEl = document.createElement('div');
      interimEl.id = `${chatInputId}-interim`;
      
      const finalEl = document.createElement('div');
      finalEl.id = `${chatInputId}-final`;
      
      transcriptContainer.appendChild(interimEl);
      transcriptContainer.appendChild(finalEl);
      if (!document.getElementById(transcriptContainer.id)) {
        document.body.appendChild(transcriptContainer);
      }
    }
    
    return () => {}; 
  }, [onSendMessage]);

  useEffect(() => {
    const textarea = textareaRef.current;
    if (textarea) {
      textarea.style.height = 'auto'; 
      const scrollHeight = textarea.scrollHeight;
      const maxHeight = 128; 
      textarea.style.height = `${Math.min(scrollHeight, maxHeight)}px`;
      textarea.style.overflowY = scrollHeight > maxHeight ? 'auto' : 'hidden';
    }
  }, [inputValue]);

  // When parent signals loading is done, reset our local sending state
  useEffect(() => {
    // console.log('[ChatInput] isLoading prop changed to:', isLoading);
    if (!isLoading) {
      // console.log('[ChatInput] isLoading is false, resetting lock. isSendingRef was:', isSendingRef.current);
      setIsSending(false);
      isSendingRef.current = false; // Reset the lock when parent is done
      // console.log('[ChatInput] Lock reset. isSendingRef is now:', isSendingRef.current);
      
    } else {
      // console.log('[ChatInput] isLoading is true, lock remains active or will be set by send action.');
    }
  }, [isLoading]);

  // Moved after variable definitions to avoid reference error

    // Voice widget should be disabled if the parent is loading OR if we've just sent a voice message
  const isVoiceWidgetDisabled = isLoading || isSending;

  // Text input should be disabled if the voice widget is, OR if the voice widget is actively listening.
  const isRealVoiceActive = voiceWidgetStatus === 'active';
  const isTickerActive = isRealVoiceActive || isDemoActive;
  const isTextInputDisabled = isVoiceWidgetDisabled || isTickerActive || voiceWidgetStatus === 'connecting';
  

  // Define callbacks after all dependencies are available
  const handleVoiceSend = useCallback((message: string) => {
    // console.log('[ChatInput] handleVoiceSend called. Current isSendingRef.current:', isSendingRef.current);
    if (isSendingRef.current) {
      // console.log('[ChatInput] Voice send blocked because isSendingRef.current is true.');
      return;
    }
    // console.log('[ChatInput] Setting lock: isSendingRef.current = true, setIsSending(true)');
    isSendingRef.current = true; // Set lock immediately
    setIsSending(true); // Set state to trigger UI re-render
    onSendMessage(message);
    
    // console.log('[ChatInput] onSendMessage called with voice message.');
  }, [onSendMessage]);

  const handleSend = useCallback(() => {
    if (isSendingRef.current) {
      // console.log('[ChatInput] Text send blocked because a message is already in flight.');
      return;
    }
    const trimmedInput = inputValue.trim();
    if (trimmedInput && !isTextInputDisabled) {
      // console.log('[ChatInput] Setting lock for text send.');
      isSendingRef.current = true;
      setIsSending(true);
      onSendMessage(trimmedInput);
      setInputValue('');
      if (textareaRef.current) {
        textareaRef.current.style.height = 'auto';
        textareaRef.current.style.overflowY = 'hidden';
        setTimeout(() => textareaRef.current?.focus(), 0);
      }
    }
  }, [inputValue, isTextInputDisabled, onSendMessage]);

  const handleKeyDown = (event: KeyboardEvent<HTMLTextAreaElement>) => {
    if (event.key === 'Enter' && !event.shiftKey) {
      event.preventDefault(); 
      handleSend();
    }
  };

  // Focus management: when input becomes enabled after bot response, focus if user interacted before
  useEffect(() => {
    if (!isTextInputDisabled && hasInteracted) {
      // small delay to let DOM update
      setTimeout(() => textareaRef.current?.focus(), 0);
    }
  }, [isTextInputDisabled, hasInteracted]);

  return (
    <div 
      className="fixed bottom-0 left-0 right-0 z-10 pt-2"
      style={{ paddingBottom: `calc(1rem + var(--safe-area-inset-bottom))` }}
    >
      {/* Further reduced height container */}
      <div className="w-[95%] max-w-md bg-black/20 backdrop-blur-xl border border-white/10 rounded-2xl shadow-lg px-4 py-1.5 mx-auto flex flex-col gap-1.5 transition-all duration-150">
        {/* Input area with SendButton - Corrected Alignment */}
        <div className="relative flex items-center w-full py-1.5">
          <textarea
            ref={textareaRef}
            value={inputValue}
            onChange={(e) => setInputValue(e.target.value) }
             onFocus={() => {
               if (!hasInteracted) setHasInteracted(true);
             }}
            onKeyDown={handleKeyDown}
            
            className="w-full bg-transparent resize-none border-none focus:outline-none focus:ring-0 pr-4 text-text placeholder:text-text/60 text-sm max-h-[96px] font-normal font-sans"
            rows={1}
            disabled={isTextInputDisabled}
            autoComplete="off"
            style={{ scrollbarWidth: 'none' }}
          />
           {/* Animated placeholder overlay */}
           <div className="absolute inset-0 flex items-center pointer-events-none">
             <AnimatePresence mode="wait">
               {!inputValue && !isTickerActive && !hasInteracted && (
                 <motion.p
                   key={`animated-placeholder-${currentPlaceholder}`}
                   initial={{ y: 5, opacity: 0 }}
                   animate={{ y: 0, opacity: 1 }}
                   exit={{ y: -15, opacity: 0 }}
                   transition={{ duration: 0.3, ease: "linear" }}
                   className="text-text/60 text-sm font-normal font-sans pl-4 text-left w-[calc(100%-2rem)] truncate"
                 >
                   {placeholders[currentPlaceholder]}
                 </motion.p>
               )}
             </AnimatePresence>
           </div>

           {/* VoiceTicker is rendered here, its internal logic handles visibility */}
          {isTickerActive && (
            <div className="absolute inset-0 pointer-events-none">
              <VoiceTicker isRecordingActive={true} recorder={voiceTickerRecorderRef} />
           </div>
          )}
        </div>

        {/* Lower row for other action buttons */}
        {/* Lower row for other action buttons */}
        <div className="flex items-center justify-between pt-2">
          <div className="flex items-center gap-4">
            <DashboardButton className="text-xs px-2.5 py-1.5" /> {/* Compact styling */}
            
            <VoiceWidget 
              onFinalTranscriptCommitted={handleVoiceSend} 
              isChatProcessing={isVoiceWidgetDisabled}
              isSendingRef={isSendingRef} 

              onStatusChange={setVoiceWidgetStatus}
              onRmsData={(rms) => {
                if (voiceTickerRecorderRef.current && voiceTickerRecorderRef.current.onResamplerData) {
                  voiceTickerRecorderRef.current.onResamplerData(rms);
                }
              }}
            />
          </div>
          <LottieSendButton
            onClick={handleSend}
            disabled={!inputValue.trim() || isTextInputDisabled}
            isSending={isSending}
            className="shadow-none"
          />
          {/* <VoiceModeToggle 
            isVoiceEnabled={voiceEnabled} 
            toggleVoiceEnabled={toggleVoice} 
            disabled={isLoading} 
            className="p-1.5" 
          /> */}
          {/* <WalkieTalkieButton 
            isVoiceEnabled={voiceEnabled} 
            isWalkieActive={isWalkieActive} 
            onClick={toggleWalkie} 
            disabled={isLoading} 
            className="p-1.5" 
          /> */}
        </div>
      </div>
    </div>
  );
});

ChatInput.displayName = 'ChatInput';

export default ChatInput;