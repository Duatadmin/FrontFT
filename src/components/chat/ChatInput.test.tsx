import { describe, test, expect, vi, beforeEach } from 'vitest';
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import ChatInput from './ChatInput'; // Adjusted import path assuming test is in same dir

// Mock hooks
const mockToggleVoice = vi.fn();
const mockToggleWalkie = vi.fn();

vi.mock('../../hooks/useVoicePlayback', () => ({
  useVoicePlayback: () => ({
    voiceEnabled: false,
    toggleVoice: mockToggleVoice,
  }),
}));

vi.mock('../../hooks/useWalkie', () => ({
  useWalkie: () => ({
    isWalkieActive: false,
    isListening: false,
    toggleWalkie: mockToggleWalkie,
    transcript: '',
    error: null,
  }),
}));

// Mock initVoiceModule from ../voice to prevent side effects if it's called
// Assuming ChatInput is in src/components/chat/ChatInput.tsx
// and voice is in src/components/voice/
// so the relative path from ChatInput.test.tsx (in src/components/chat/)
// to src/components/voice/ would be '../voice'
vi.mock('../voice', () => ({
  initVoiceModule: vi.fn(() => Promise.resolve()),
}));

// Mock react-router-dom for the Link component in DashboardButton
vi.mock('react-router-dom', () => ({
  ...vi.importActual('react-router-dom'), // import and retain default behavior
  Link: ({ children, to }: { children: React.ReactNode, to: string }) => <a href={to}>{children}</a>,
}));


describe('ChatInput Component', () => {
  const mockOnSendMessage = vi.fn();
  const defaultProps = {
    onSendMessage: mockOnSendMessage,
    isLoading: false,
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('renders correctly and matches snapshot', () => {
    const { container } = render(<ChatInput {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  test('input field updates value on change', () => {
    render(<ChatInput {...defaultProps} />);
    const textarea = screen.getByPlaceholderText('Ask anything...') as HTMLTextAreaElement;
    fireEvent.change(textarea, { target: { value: 'Hello there' } });
    expect(textarea.value).toBe('Hello there');
  });

  test('calls onSendMessage when SendButton is clicked with input', () => {
    render(<ChatInput {...defaultProps} />);
    const textarea = screen.getByPlaceholderText('Ask anything...') as HTMLTextAreaElement;
    const sendButton = screen.getByLabelText('Send message');

    fireEvent.change(textarea, { target: { value: 'Test message' } });
    fireEvent.click(sendButton);

    expect(mockOnSendMessage).toHaveBeenCalledWith('Test message');
    expect((textarea as HTMLTextAreaElement).value).toBe(''); // Assuming input clears on send
  });

  test('SendButton is disabled when input is empty', () => {
    render(<ChatInput {...defaultProps} />);
    const sendButton = screen.getByLabelText('Send message');
    expect(sendButton).toBeDisabled();
  });

  test('SendButton is enabled when input is not empty', () => {
    render(<ChatInput {...defaultProps} />);
    const textarea = screen.getByPlaceholderText('Ask anything...') as HTMLTextAreaElement;
    fireEvent.change(textarea, { target: { value: 'Not empty' } });
    const sendButton = screen.getByLabelText('Send message');
    expect(sendButton).not.toBeDisabled();
  });

  test('calls toggleVoice when VoiceModeToggle is clicked', () => {
    render(<ChatInput {...defaultProps} />);
    const voiceToggle = screen.getByLabelText('Enable voice mode'); // Initial label
    fireEvent.click(voiceToggle);
    expect(mockToggleVoice).toHaveBeenCalledTimes(1);
  });

  test('calls toggleWalkie when WalkieTalkieButton is clicked', () => {
    render(<ChatInput {...defaultProps} />);
    const walkieButton = screen.getByLabelText('Start voice input'); // Initial label
    fireEvent.click(walkieButton);
    expect(mockToggleWalkie).toHaveBeenCalledTimes(1);
  });

  test('DashboardButton renders and links to /dashboard', () => {
    render(<ChatInput {...defaultProps} />);
    const dashboardButton = screen.getByLabelText('Open Dashboard') as HTMLAnchorElement;
    expect(dashboardButton).toBeInTheDocument();
    expect(dashboardButton.getAttribute('href')).toBe('/dashboard');
  });

});
