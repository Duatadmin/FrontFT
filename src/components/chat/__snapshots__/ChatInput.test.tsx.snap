// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatInput Component > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="sticky bottom-0 z-10 pt-2 pb-2"
  >
    <div
      class="w-[min(95vw,525px)] min-w-[min(95vw,525px)] bg-botBubble shadow-2xl rounded-3xl px-4 py-1.5 mx-auto flex flex-col gap-1.5"
    >
      <div
        class="relative flex items-center w-full"
      >
        <textarea
          autocomplete="off"
          class="w-full bg-transparent resize-none border-none focus:outline-none focus:ring-0 pr-12 py-1.5 text-text placeholder:text-text/60 text-sm max-h-[96px] font-normal font-sans"
          placeholder="Ask anything..."
          rows="1"
          style="scrollbar-width: none; height: 0px; overflow-y: hidden;"
        />
        <button
          aria-label="Send message"
          class="p-2 rounded-full text-slate-400 hover:text-slate-100 hover:bg-slate-800 active:bg-slate-700 transition-colors duration-150 ease-in-out focus:outline-none focus-visible:ring-2 focus-visible:ring-sky-500 focus-visible:ring-offset-2 focus-visible:ring-offset-slate-900 disabled:opacity-50 disabled:cursor-not-allowed absolute right-0 top-1/2 -translate-y-1/2 h-[38px] w-[38px]"
          disabled=""
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-arrow-up-right"
            fill="none"
            height="22"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="22"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 7h10v10"
            />
            <path
              d="M7 17 17 7"
            />
          </svg>
        </button>
      </div>
      <div
        class="flex items-center justify-start gap-2 pt-2"
      >
        <a
          href="/dashboard"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-layout-dashboard"
            fill="none"
            height="18"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="9"
              rx="1"
              width="7"
              x="3"
              y="3"
            />
            <rect
              height="5"
              rx="1"
              width="7"
              x="14"
              y="3"
            />
            <rect
              height="9"
              rx="1"
              width="7"
              x="14"
              y="12"
            />
            <rect
              height="5"
              rx="1"
              width="7"
              x="3"
              y="16"
            />
          </svg>
          <span>
            Dashboard
          </span>
        </a>
         
        <button
          aria-label="Enable voice mode"
          aria-pressed="false"
          class="p-2 rounded-full text-slate-400 hover:text-slate-100 hover:bg-slate-800 active:bg-slate-700 transition-colors duration-150 ease-in-out focus:outline-none focus-visible:ring-2 focus-visible:ring-sky-500 focus-visible:ring-offset-2 focus-visible:ring-offset-slate-900 disabled:opacity-50 disabled:cursor-not-allowed p-1.5"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-volume-x"
            fill="none"
            height="22"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="22"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
            />
            <line
              x1="22"
              x2="16"
              y1="9"
              y2="15"
            />
            <line
              x1="16"
              x2="22"
              y1="9"
              y2="15"
            />
          </svg>
        </button>
        <button
          aria-label="Start voice input"
          aria-pressed="false"
          class="p-2 rounded-full hover:bg-slate-800 active:bg-slate-700 transition-colors duration-150 ease-in-out focus:outline-none focus-visible:ring-2 focus-visible:ring-sky-500 focus-visible:ring-offset-2 focus-visible:ring-offset-slate-900 disabled:opacity-50 disabled:cursor-not-allowed text-slate-400 p-1.5"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-mic"
            fill="none"
            height="22"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="22"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 2a3 3 0 0 0-3 3v7a3 3 0 0 0 6 0V5a3 3 0 0 0-3-3Z"
            />
            <path
              d="M19 10v2a7 7 0 0 1-14 0v-2"
            />
            <line
              x1="12"
              x2="12"
              y1="19"
              y2="22"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
